lib/src/logger/deprecation_handling.dart

```diff
@@ -0,0 +1,103 @@
+// Copyright 2018 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'package:collection/collection.dart';
+import 'package:source_span/source_span.dart';
+import 'package:stack_trace/stack_trace.dart';
+
+import '../deprecation.dart';
+import '../exception.dart';
+import '../logger.dart';
+
+/// The maximum number of repetitions of the same warning
+/// [DeprecationHandlingLogger] will emit before hiding the rest.
+const _maxRepetitions = 5;
+
+/// A logger that wraps an inner logger to have special handling for
+/// deprecation warnings.
+class DeprecationHandlingLogger implements Logger {
+  /// A map of how many times each deprecation has been emitted by this logger.
+  final _warningCounts = <Deprecation, int>{};
+
+  final Logger _inner;
+
+  /// Deprecation warnings of one of these types will cause an error to be
+  /// thrown.
+  ///
+  /// Future deprecations in this list will still cause an error even if they
+  /// are not also in [futureDeprecations].
+  final Set<Deprecation> fatalDeprecations;
+
+  /// Future deprecations that the user has explicitly opted into.
+  final Set<Deprecation> futureDeprecations;
+
+  /// Whether repetitions of the same warning should be limited to no more than
+  /// [_maxRepetitions].
+  final bool limitRepetition;
+
+  DeprecationHandlingLogger(this._inner,
+      {required this.fatalDeprecations,
+      required this.futureDeprecations,
+      this.limitRepetition = true});
+
+  void warn(String message,
+      {FileSpan? span, Trace? trace, bool deprecation = false}) {
+    _inner.warn(message, span: span, trace: trace, deprecation: deprecation);
+  }
+
+  /// Processes a deprecation warning.
+  ///
+  /// If [deprecation] is in [fatalDeprecations], this shows an error.
+  ///
+  /// If it's a future deprecation that hasn't been opted into or its a
+  /// deprecation that's already been warned for [_maxReptitions] times and
+  /// [limitRepetitions] is true, the warning is dropped.
+  ///
+  /// Otherwise, this is passed on to [warn].
+  void handleDeprecationWarning(
+      Deprecation deprecation, String message, FileSpan? span, Trace? trace) {
+    if (fatalDeprecations.contains(deprecation)) {
+      message += '\n\nThis is only an error because of '
+          '--fatal-deprecation=$deprecation.\n'
```

Comment:
The CLI flag format may not be accurate for library users.

Reply:
Updated this to be more generic. (Or would it make more sense to try and detect CLI vs API use and update the message accordingly?)

Reply:
I think making it generic is fine.